### PR TITLE
*: use schema diff to reduce memory allocation

### PR DESCRIPTION
**Pull request by coocood**
When schema version has changed, we don't need to reload all schemas, only
modified table need to be reloaded, this will reduce the pressure on Go garbage collector.

Also removed used map in *inforSchema.

### CODE REVIEW

**Comment by ngaut** [comment]
PTAL @zimulala @tiancaiamao @disksing

File: privilege/privileges/privileges_test.go [hunk 1]
```diff
@@ -56,9 +56,11 @@ type testPrivilegeSuite struct {
 func (s *testPrivilegeSuite) SetUpSuit(c *C) {
 	logLevel := os.Getenv("log_level")
 	log.SetLevelByString(logLevel)
+	log.SetLevelByString("error")
 }
 
 func (s *testPrivilegeSuite) SetUpTest(c *C) {
+	fmt.Println("SETUPTEST")
```
Comment (ngaut) [inline privilege/privileges/privileges_test.go]: Remove debug messages.

File: privilege/privileges/privileges_test.go [hunk 2]
```diff
@@ -71,7 +73,9 @@ func (s *testPrivilegeSuite) SetUpTest(c *C) {
 	mustExec(c, se, s.createDBSQL)
 	mustExec(c, se, s.createDB1SQL) // create database test1
 	mustExec(c, se, s.useDBSQL)
-	mustExec(c, se, s.createTableSQL)
+	_, err := se.Execute(s.createTableSQL)
+	log.Error(err)
```
Comment (ngaut) [inline privilege/privileges/privileges_test.go]: Remove debug messages.

File: privilege/privileges/privileges_test.go [hunk 3]
```diff
@@ -87,9 +91,11 @@ func (s *testPrivilegeSuite) SetUpTest(c *C) {
 }
 
 func (s *testPrivilegeSuite) TearDownTest(c *C) {
+	fmt.Println("TEARDOWNTEST")
```
Comment (ngaut) [inline privilege/privileges/privileges_test.go]: Remove debug messages.

File: domain/domain.go [hunk 4]
```diff
@@ -86,9 +107,44 @@ func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion in
...
+	return schemas, nil
+}
+
+// tryLoadSchemaDiffs tries to only load latest schema changes.
+func (do *Domain) tryLoadSchemaDiffs(m *meta.Meta, usedVersion, newVersion int64) (bool, error) {
+	if usedVersion == 0 || newVersion-usedVersion > 100 {
+		// If there isn't any used version, or used version is too old, we do full load.
+		return false, nil
+	}
+	if usedVersion > newVersion {
```
Comment (ngaut) [inline domain/domain.go]: Is is possible that usedVersion larger than newVersion ?

File: domain/domain.go [hunk 5]
```diff
@@ -86,9 +107,44 @@ func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion in
...
 		}
 	}
-	log.Infof("[ddl] loadInfoSchema %d", latestSchemaVersion)
-	err = handle.Set(schemas, latestSchemaVersion)
-	return errors.Trace(err)
+	return schemas, nil
+}
+
+// tryLoadSchemaDiffs tries to only load latest schema changes.
+func (do *Domain) tryLoadSchemaDiffs(m *meta.Meta, usedVersion, newVersion int64) (bool, error) {
```
Comment (ngaut) [inline domain/domain.go]: Please add comments to explain bool return value.

File: infoschema/builder.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+	b.is.schemaMetaVersion = diff.Version
+	if diff.Type == model.ActionCreateSchema {
+		return b.applyCreateSchema(m, diff)
+	} else if diff.Type == model.ActionDropSchema {
+		b.applyDropSchema(diff.SchemaID)
+		return nil
+	}
+	di, ok := b.is.schemas[diff.SchemaID]
+	if !ok {
+		return ErrDatabaseNotExists
```
Comment (ngaut) [inline infoschema/builder.go]: There must be some serious bug if we go into the branch. Maybe we should log fatal message.

File: meta/meta.go [hunk 7]
```diff
@@ -735,6 +736,36 @@ func (m *Meta) GetTableStats(tableID int64) (*statistics.TablePB, error) {
 	return tpb, nil
 }
 
+func (m *Meta) schemaDiffKey(schemaVersion int64) []byte {
+	return []byte(fmt.Sprintf("%s:%d", mSchemaDiffPrefix, schemaVersion))
+}
+
+// GetSchemaDiff gets the modification information on a given schema version.
```
Comment (zimulala) [inline meta/meta.go]: s/modification/modified

File: meta/meta.go [hunk 8]
```diff
@@ -735,6 +736,36 @@ func (m *Meta) GetTableStats(tableID int64) (*statistics.TablePB, error) {
...
+func (m *Meta) GetSchemaDiff(schemaVersion int64) (*model.SchemaDiff, error) {
```
Comment (zimulala) [inline meta/meta.go]: It need tests.

File: domain/domain.go [hunk 9]
```diff
@@ -60,10 +60,31 @@ func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion in
...
+	schemas, err := do.getAllSchemasWithTablesFromMeta(m)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	newISBuilder, err := infoschema.NewBuilder(handle).InitWithDBInfos(schemas, latestSchemaVersion)
 	if err != nil {
 		return errors.Trace(err)
 	}
+	log.Infof("[ddl] loadInfoSchema %d", latestSchemaVersion)
```
Comment (zimulala) [inline domain/domain.go]: Update the log and add some logs at line68.

File: infoschema/builder.go [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+// ApplyDiff applies SchemaDiff to the new InfoSchema.
+func (b *Builder) ApplyDiff(m *meta.Meta, diff *model.SchemaDiff) error {
+	b.is.schemaMetaVersion = diff.Version
+	if diff.Type == model.ActionCreateSchema {
+		return b.applyCreateSchema(m, diff)
+	} else if diff.Type == model.ActionDropSchema {
+		b.applyDropSchema(diff.SchemaID)
+		return nil
+	}
+	di, ok := b.is.schemas[diff.SchemaID]
```
Comment (ngaut) [inline infoschema/builder.go]: s/di/roDBInfo/

File: infoschema/builder.go [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	// The old DBInfo still holds a reference to old table info, we need to update it.
+	// As old DBInfo should be read only, we clone a new DBInfo.
+	newDbInfo := new(model.DBInfo)
+	*newDbInfo = *di
+	newDbInfo.Tables = make([]*model.TableInfo, 0, len(di.Tables))
+	if newTableID != 0 {
```
Comment (ngaut) [inline infoschema/builder.go]: Define 0 as constant.

File: meta/meta.go [hunk 12]
```diff
@@ -735,6 +736,36 @@ func (m *Meta) GetTableStats(tableID int64) (*statistics.TablePB, error) {
...
+	diffKey := m.schemaDiffKey(schemaVersion)
+	data, err := m.txn.Get(diffKey)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(data) == 0 {
+		return nil, nil
+	}
+	diff := new(model.SchemaDiff)
```
Comment (ngaut) [inline meta/meta.go]: var diff model.SchemaDiff

File: domain/domain.go [hunk 13]
```diff
@@ -86,9 +107,44 @@ func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion in
...
 	}
-	log.Infof("[ddl] loadInfoSchema %d", latestSchemaVersion)
-	err = handle.Set(schemas, latestSchemaVersion)
-	return errors.Trace(err)
+	return schemas, nil
+}
+
+// tryLoadSchemaDiffs tries to only load latest schema changes.
+func (do *Domain) tryLoadSchemaDiffs(m *meta.Meta, usedVersion, newVersion int64) (bool, error) {
+	if usedVersion == 0 || newVersion-usedVersion > 100 {
```
Comment (hanfei1991) [inline domain/domain.go]: Can we set 100 as a const ?

File: infoschema/builder.go (hunk 11 above)
Comment (hanfei1991) [inline infoschema/builder.go]: when the new TableID is equal to 0 ? add comments here.

File: meta/meta.go (hunk 7 above)
Comment (coocood) [inline meta/meta.go]: I think modification is fine.

File: infoschema/builder.go (hunk 6 above)
Comment (coocood) [inline infoschema/builder.go]: If that happens, we can fall back to full load, we don't have to `log.Fatal`.

File: meta/meta.go (hunk 12 above)
Comment (coocood) [inline meta/meta.go]: We need to return a pointer type.

**Comment by coocood** [comment]
@zimulala @ngaut @hanfei1991 PTAL

File: domain/domain.go [hunk 14]
```diff
@@ -86,9 +109,49 @@ func (do *Domain) loadInfoSchema(handle *infoschema.Handle, usedSchemaVersion in
...
+	return schemas, nil
+}
+
+const maxNumberOfDiffsToLoad = 100
+
+// tryLoadSchemaDiffs tries to only load latest schema changes.
+// Returns true if the schema is loaded successfully.
+// Returns false if the schema can not be loaded by schema diff, then we need to do full load.
+func (do *Domain) tryLoadSchemaDiffs(m *meta.Meta, usedVersion, newVersion int64) (bool, error) {
+	if usedVersion == 0 || newVersion-usedVersion > maxNumberOfDiffsToLoad {
```
Comment (ngaut) [inline domain/domain.go]: define 0 as constant.

File: infoschema/builder.go [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+		// All types except DropTable.
+		err := b.applyCreateTable(m, roDBInfo, newTableID, alloc)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	// The old DBInfo still holds a reference to old table info, we need to update it.
+	// As old DBInfo should be read only, we clone a new DBInfo.
+	newDbInfo := new(model.DBInfo)
```
Comment (ngaut) [inline infoschema/builder.go]: Extract to a new function, cloneDBInfo().

**Comment by hanfei1991** [comment]
LGTM

File: infoschema/builder.go [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+		return
+	}
+	delete(b.is.schemas, di.ID)
+	delete(b.is.schemaNameToID, di.Name.L)
+	for _, tbl := range di.Tables {
+		b.applyDropTable(di.Name.L, tbl.ID)
+	}
+}
+
+func (b *Builder) applyCreateTable(m *meta.Meta, di *model.DBInfo, tableID int64, alloc autoid.Allocator) error {
```
Comment (zimulala) [inline infoschema/builder.go]: s/di/roDBInfo ?

**Comment by ngaut** [comment]
LGTM

**Pull request merged**